src/FileFormats/NL/NL.jl

```diff
@@ -0,0 +1,788 @@
+    # is _strictly_ greater than the number of nonlinear variables in the
+    # constraints only. Quoting:
+    #
+    #   For all versions, the first nlvc variables appear nonlinearly in at
+    #   least one constraint. If nlvo > nlvc, the first nlvc variables may or
+    #   may not appear nonlinearly in an objective, but the next nlvo – nlvc
+    #   variables do appear nonlinearly in at least one objective. Otherwise
+    #   all of the first nlvo variables appear nonlinearly in an objective.
+    #
+    # However, even this is slightly incorrect, because I think it should read
+    # "For all versions, the first nlvb variables appear nonlinearly." The "nlvo
+    # - nlvc" part is also clearly incorrect, and should probably read "nlvo -
+    # nlvb."
+    #
+    # It's a bit confusing, so here is the relevant code from Couenne:
+    # https://github.com/coin-or/Couenne/blob/683c5b305d78a009d59268a4bca01e0ad75ebf02/src/readnl/readnl.cpp#L76-L87
+    #
+    # They interpret this paragraph to mean the switch on nlvo > nlvc determines
+    # whether the next block of variables are the ones that appear in the
+    # objective only, or the constraints only.
+    #
+    # That makes sense as a design choice, because you can read them in two
+    # contiguous blocks.
+    #
+    # Essentially, what all this means is if !(nlvo > nlvc), then swap 3-4 for
+    # 5-6 in the variable order.
+    nlvc = length(types[3]) + length(types[4])
+    nlvo = length(types[5]) + length(types[6])
+    order_i = if nlvo > nlvc
+        [1, 2, 3, 4, 5, 6, 7, 8, 9]
+    else
+        [1, 2, 5, 6, 3, 4, 7, 8, 9]
+    end
+    # Now we can order the variables.
+    n = 0
+    for i in order_i
+        # Since variables come from a dictionary, there may be differences in
+        # the order depending on platform and Julia version. Sort by creation
+        # time for consistency.
+        for x in sort!(types[i]; by = y -> y.value)
+            dest.x[x].order = n
+            dest.order[n+1] = x
+            n += 1
+        end
+    end
+    return mapping
+end
+
+_set_to_bounds(set::MOI.Interval) = (0, set.lower, set.upper)
+_set_to_bounds(set::MOI.LessThan) = (1, -Inf, set.upper)
+_set_to_bounds(set::MOI.GreaterThan) = (2, set.lower, Inf)
+_set_to_bounds(set::MOI.EqualTo) = (4, set.value, set.value)
+
+function _process_constraint(dest::Model, model, F, S, mapping)
+    for ci in MOI.get(model, MOI.ListOfConstraintIndices{F,S}())
+        f = MOI.get(model, MOI.ConstraintFunction(), ci)
+        s = MOI.get(model, MOI.ConstraintSet(), ci)
+        op, l, u = _set_to_bounds(s)
+        con = _NLConstraint(l, u, op, _NLExpr(f))
+        if isempty(con.expr.linear_terms) && isempty(con.expr.nonlinear_terms)
+            if l <= con.expr.constant <= u
+                continue
+            else
+                error(
+                    "Malformed constraint. There are no variables and the " *
+                    "bounds don't make sense.",
```

Comment:
"Malformed constraint $ci ... : the function constant `$con.expr.constant` is not in the interval [`$l`, `$u`]"